File: lib/api/adminCorrections.ts
/**
 * Admin Corrections API client — Human-in-the-Loop review workflow.
 *
 * Backend routes (see `backend/app/api/routes/admin_corrections.py`):
 *
 *   GET  /api/admin/corrections/queue
 *   GET  /api/admin/corrections/traces/{kind}/{id}
 *   POST /api/admin/corrections
 *   GET  /api/admin/corrections
 *   GET  /api/admin/corrections/{id}
 *   POST /api/admin/corrections/{id}/review
 *   GET  /api/admin/corrections/export.jsonl
 */
import { apiUrl } from "@/lib/api/base";

async function authFetch<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, { credentials: "include", ...init });
  if (!response.ok) {
    const payload = await response
      .json()
      .catch(() => ({ detail: "Request failed" }));
    throw new Error(payload.detail ?? "Request failed");
  }
  return response.json();
}

// ── Types ────────────────────────────────────────────────────────────────

export type CorrectionTargetKind = "assistant" | "vision";
export type CorrectionStatus = "pending" | "applied" | "rejected";

export interface TraceQueueItem {
  trace_id: string;
  target_kind: CorrectionTargetKind;
  user_id?: string | null;
  user_name?: string | null;
  user_email?: string | null;
  created_at: string;
  question_preview?: string | null;
  answer_preview?: string | null;
  image_thumbnail_url?: string | null;
  confidence?: number | null;
  has_existing_correction: boolean;
}

export interface TraceQueueResponse {
  items: TraceQueueItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface TraceCitation {
  document_id?: string | null;
  document_title?: string | null;
  page?: number | null;
  snippet?: string | null;
  score?: number | null;
}

export interface AgentStep {
  step_index: number;
  step_type: string;
  input_json?: Record<string, unknown> | null;
  output_json?: Record<string, unknown> | null;
  elapsed_ms?: number | null;
}

export interface AdminCorrection {
  id: string;
  target_kind: CorrectionTargetKind;
  assistant_trace_id?: string | null;
  vision_trace_id?: string | null;

  admin_user_id?: string | null;
  admin_name?: string | null;

  original_text?: string | null;
  original_payload?: Record<string, unknown> | null;

  corrected_text: string;
  corrected_payload?: Record<string, unknown> | null;
  rationale?: string | null;
  concept_tags?: string[] | null;

  status: CorrectionStatus;
  reviewed_by_user_id?: string | null;
  reviewed_by_name?: string | null;
  reviewed_at?: string | null;
  review_notes?: string | null;

  created_at: string;
  updated_at: string;
}

export interface TraceDetail {
  trace_id: string;
  target_kind: CorrectionTargetKind;
  user_id?: string | null;
  user_name?: string | null;
  created_at: string;

  question?: string | null;
  answer: string;
  answer_payload?: Record<string, unknown> | null;
  citations: TraceCitation[];

  image_url?: string | null;
  heatmap_url?: string | null;
  findings?: Record<string, unknown>[] | null;

  confidence?: number | null;
  faithfulness_score?: number | null;
  agent_steps: AgentStep[];

  existing_correction?: AdminCorrection | null;
}

export interface AdminCorrectionList {
  items: AdminCorrection[];
  total: number;
  page: number;
  page_size: number;
}

export interface AdminCorrectionCreatePayload {
  target_kind: CorrectionTargetKind;
  assistant_trace_id?: string;
  vision_trace_id?: string;
  corrected_text: string;
  corrected_payload?: Record<string, unknown>;
  rationale?: string;
  concept_tags?: string[];
}

export interface AdminCorrectionReviewPayload {
  decision: "apply" | "reject";
  review_notes?: string;
}

// ── API ──────────────────────────────────────────────────────────────────

export async function getCorrectionQueue(params?: {
  targetKind?: CorrectionTargetKind;
  onlyUncorrected?: boolean;
  page?: number;
  pageSize?: number;
}): Promise<TraceQueueResponse> {
  const qs = new URLSearchParams();
  if (params?.targetKind) qs.set("target_kind", params.targetKind);
  if (params?.onlyUncorrected) qs.set("only_uncorrected", "true");
  if (params?.page) qs.set("page", String(params.page));
  if (params?.pageSize) qs.set("page_size", String(params.pageSize));
  return authFetch<TraceQueueResponse>(
    apiUrl(`/admin/corrections/queue?${qs.toString()}`)
  );
}

export async function getTraceDetail(
  kind: CorrectionTargetKind,
  traceId: string
): Promise<TraceDetail> {
  return authFetch<TraceDetail>(
    apiUrl(`/admin/corrections/traces/${kind}/${traceId}`)
  );
}

export async function submitCorrection(
  payload: AdminCorrectionCreatePayload
): Promise<AdminCorrection> {
  return authFetch<AdminCorrection>(apiUrl(`/admin/corrections`), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

export async function listCorrections(params?: {
  status?: CorrectionStatus;
  targetKind?: CorrectionTargetKind;
  page?: number;
  pageSize?: number;
}): Promise<AdminCorrectionList> {
  const qs = new URLSearchParams();
  if (params?.status) qs.set("status", params.status);
  if (params?.targetKind) qs.set("target_kind", params.targetKind);
  if (params?.page) qs.set("page", String(params.page));
  if (params?.pageSize) qs.set("page_size", String(params.pageSize));
  return authFetch<AdminCorrectionList>(
    apiUrl(`/admin/corrections?${qs.toString()}`)
  );
}

export async function getCorrection(id: string): Promise<AdminCorrection> {
  return authFetch<AdminCorrection>(apiUrl(`/admin/corrections/${id}`));
}

export async function reviewCorrection(
  id: string,
  payload: AdminCorrectionReviewPayload
): Promise<AdminCorrection> {
  return authFetch<AdminCorrection>(apiUrl(`/admin/corrections/${id}/review`), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

export function exportCorrectionsJsonlUrl(status: "pending" | "applied" | "rejected" | "all" = "applied"): string {
  return apiUrl(`/admin/corrections/export.jsonl?status=${status}`);
}

File: backend/app/schemas/admin_corrections.py
"""
Pydantic schemas for the admin Human-in-the-Loop (HIL) correction workflow.

These schemas describe the JSON contracts of `/admin/corrections/*` routes
that allow the three production admins to:

  * inspect recent assistant / vision AI traces (the "queue"),
  * submit an authoritative correction for a single trace,
  * review pending corrections submitted by another admin,
  * apply / reject corrections (lifecycle transitions),
  * and export the full corpus as JSONL for downstream fine-tuning /
    distillation.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


# ── Queue items ────────────────────────────────────────────────────────────

class TraceQueueItemOut(BaseModel):
    """A single AI trace surfaced in the admin review queue."""

    trace_id: str
    target_kind: Literal["assistant", "vision"]
    user_id: str | None = None
    user_name: str | None = None
    user_email: str | None = None
    created_at: datetime

    # Display-only summary fields (truncated server-side).
    question_preview: str | None = None
    answer_preview: str | None = None
    image_thumbnail_url: str | None = None  # vision only

    # Existing AI confidence / faithfulness if available.
    confidence: float | None = None
    has_existing_correction: bool = False


class TraceQueueOut(BaseModel):
    items: list[TraceQueueItemOut]
    total: int
    page: int
    page_size: int


# ── Trace detail (used by the Trace Inspector pane) ────────────────────────

class TraceCitationOut(BaseModel):
    document_id: str | None = None
    document_title: str | None = None
    page: int | None = None
    snippet: str | None = None
    score: float | None = None


class AgentStepOut(BaseModel):
    step_index: int
    step_type: str
    input_json: dict | None = None
    output_json: dict | None = None
    elapsed_ms: int | None = None


class TraceDetailOut(BaseModel):
    trace_id: str
    target_kind: Literal["assistant", "vision"]
    user_id: str | None = None
    user_name: str | None = None
    created_at: datetime

    # Question + answer.
    question: str | None = None
    answer: str
    answer_payload: dict | None = None  # structured (e.g. flashcards/quizzes)
    citations: list[TraceCitationOut] = []

    # Vision-only.
    image_url: str | None = None
    heatmap_url: str | None = None
    findings: list[dict] | None = None

    # Diagnostics.
    confidence: float | None = None
    faithfulness_score: float | None = None
    agent_steps: list[AgentStepOut] = []

    # Existing correction (if any).
    existing_correction: AdminCorrectionOut | None = None


# ── Submit / review corrections ────────────────────────────────────────────

class AdminCorrectionCreate(BaseModel):
    """Payload to create a new correction for a trace."""

    target_kind: Literal["assistant", "vision"]
    assistant_trace_id: str | None = None
    vision_trace_id: str | None = None

    corrected_text: str = Field(..., min_length=1)
    corrected_payload: dict | None = None
    rationale: str | None = None
    concept_tags: list[str] | None = None


class AdminCorrectionReview(BaseModel):
    """Payload for a second-admin review (apply / reject)."""

    decision: Literal["apply", "reject"]
    review_notes: str | None = None


class AdminCorrectionOut(BaseModel):
    id: str
    target_kind: Literal["assistant", "vision"]
    assistant_trace_id: str | None = None
    vision_trace_id: str | None = None

    admin_user_id: str | None = None
    admin_name: str | None = None

    original_text: str | None = None
    original_payload: dict | None = None

    corrected_text: str
    corrected_payload: dict | None = None
    rationale: str | None = None
    concept_tags: list[str] | None = None

    status: Literal["pending", "applied", "rejected"]
    reviewed_by_user_id: str | None = None
    reviewed_by_name: str | None = None
    reviewed_at: datetime | None = None
    review_notes: str | None = None

    created_at: datetime
    updated_at: datetime


class AdminCorrectionListOut(BaseModel):
    items: list[AdminCorrectionOut]
    total: int
    page: int
    page_size: int


# Resolve forward refs introduced by the existing-correction nesting on
# TraceDetailOut.
TraceDetailOut.model_rebuild()


# ── Export ────────────────────────────────────────────────────────────────

class AdminCorrectionExportRow(BaseModel):
    """One JSONL row in the export feed (consumed by fine-tuning jobs)."""

    correction_id: str
    target_kind: Literal["assistant", "vision"]
    question: str | None = None
    original_answer: str | None = None
    corrected_answer: str
    rationale: str | None = None
    concept_tags: list[str] = []
    citations: list[Any] = []
    image_url: str | None = None
    created_at: datetime
